flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1264,6 +1278,326 @@ def test_all_transitions(self, desired_state, discovered_state):
+                ),
+            ],
+            local_datasets=[],
+            expected_datasets=[
+                DesiredDataset(
+                    state=DatasetStates.DELETED,
+                    dataset_id=ScenarioMixin.DATASET_ID,
+                    metadata=ScenarioMixin.METADATA,
+                ),
+            ],
+        )
+
+    @given(
+        expected_size=integers(min_value=0),
+    )
+    def test_leased_mounted_manifestation(self, expected_size):
+        """
+        If there is a lease for a mounted dataset present on node, there is a
+        corresponding desired dataset that has a state of ``MOUNTED`` even if
+        the configuration of the node doesn't mention the dataset.
+        """
+        assert_desired_datasets(
+            self, self.deployer,
+            desired_manifestations=[],
+            local_datasets=[
+                ScenarioMixin.MOUNTED_DISCOVERED_DATASET.transform(
+                    ['maximum_size'], expected_size,
+                ),
+            ],
+            expected_datasets=[
+                ScenarioMixin.MOUNTED_DESIRED_DATASET.transform(
+                    ['maximum_size'], expected_size,
+                ),
+            ],
+            leases=Leases().acquire(
+                now=datetime.now(tz=UTC),
+                dataset_id=ScenarioMixin.DATASET_ID,
+                node_id=self.deployer.node_uuid,
+            )
+        )
+
+    @given(
+        local_dataset=DISCOVERED_DATASET_STRATEGY.filter(
+            lambda dataset: dataset.state != DatasetStates.MOUNTED,
+        ),
+    )
+    def test_leased_not_mounted(self, local_dataset):
+        """
+        If there is a lease for a dataset that isn't mounted on the node and
+        the dataset isn't requested in the configuration of the node, there is
+        not a corresponding desired dataset.
+        """
+        assert_desired_datasets(
+            self, self.deployer,
+            desired_manifestations=[],
+            local_datasets=[local_dataset],
+            expected_datasets=[],
+            leases=Leases().acquire(
+                now=datetime.now(tz=UTC),
+                dataset_id=local_dataset.dataset_id,
+                node_id=self.deployer.node_uuid,
+            )
+        )
+
+    def test_lease_elsewhere(self):
+        """
+        If there is a lease for a dataset on another node, there isn't a
+        corresponding desired dataset.
+        """
+        assert_desired_datasets(
+            self, self.deployer,
+            local_datasets=[
+                ScenarioMixin.MOUNTED_DISCOVERED_DATASET,
+            ],
+            expected_datasets=[],
+            leases=Leases().acquire(
+                now=datetime.now(tz=UTC),
+                dataset_id=ScenarioMixin.DATASET_ID,
+                node_id=uuid4(),
+            )
+        )
+
+    def test_application_mounted_manifestation(self):
+        """
+        If there is an application with attached volume, there is a
```

Comment:
I don't *think* this is a change in behavior. This logic triggers if
- there is a lease on this node or a running (not configured) application using it
*and*
- we discovered the dataset is already mounted.